### DISCUSSION TITLE
How to decorate output with position?

### DISCUSSION THREAD

**Discussion by Iltotore** [j-mie6/Parsley#133] [Q&A] [answered]
Hello.

I am interested by the project and comparing it to similar ones like Scala Parsing Combinators, ZIO parse etc... In most of them, you can give the input position to the generated output. For example, Scala Parsing Combinators has the `positioned` combinator but I didn't find any equivalent in Parsley's wiki.

Does Parsley have such feature? If yes, is there a wiki page about its usage?

**Comment by j-mie6** [comment, marked as answer]
`Parsley.line`, `Parsley.col` and `Parsley.pos` 🙂 (see https://javadoc.io/static/com.github.j-mie6/parsley_2.13/4.0.0/parsley/Parsley$.html#pos:parsley.Parsley[(Int,Int)])

**Reply by j-mie6** [reply]
The Wiki does have a page that includes them, namely the one about the [Parser Bridge pattern](https://github.com/j-mie6/Parsley/wiki/The-Parser-Bridge-Pattern), as the example there is annotating an AST node with position information without compromising the parser. This pattern would be the equivalent of the `positioned` in a way, you'd create the bridges that automatically add the position into the data as its constructed, as opposed to mutating it afterwards. But a `positioned` combinator is fairly easy to build using `pos` as a user. This might be something to consider adding to the library, although I feel like the Parser Bridge pattern is the more idiomatic (and immutable) solution: I'd welcome your thoughts!

**Reply by j-mie6** [reply]
```scala
def positioned[T <: Positional](p: =>Parsley[T]): Parsley[T] = {
    lift2[(Int, Int), T, T]((pos, x) => {x.setPos(pos); x}, pos, p)
    // or, with nicer syntax but still preserving laziness (with `~`)
    (pos, ~p).zipped { (pos, x) =>
        x.setPos(pos)
        x
    }
}
```

**Reply by Iltotore** [reply]
This is exactly what I am looking for. Thank you!
